### ISSUE TITLE
.persist won't update entity when it contains OneToOne relationship?

### ISSUE THREAD

**Issue by codgician** [typeorm/typeorm#836]
I'm storing user profile and user credentials separately in two entities: "User" and "UserCredential". They look like this:

``` Typescript
/* User */

@Entity()
export class User {

  @PrimaryGeneratedColumn()
  @Index()
  public id: number;

  @Column('varchar', {
    unique: true,
  })
  @Index()
  public email: string;

  @Column('varchar', {
    unique: true,
  })
  @Index()
  public username: string;

  @Column('tinyint', {
    default: 0,
  })
  @Index()
  public privilege: number;

  @CreateDateColumn()
  public createdAt: string;

  @UpdateDateColumn()
  public updatedAt: string;

}

/* UserCredential */

@Entity()
export class UserCredential {

  @OneToOne(() => User, (user) => user.id, {
    primary: true,
    cascadeAll: true,
  })
  @JoinColumn({
    name: 'id',
    referencedColumnName: 'id',
  })
  @Index()
  public user: User;

  @CreateDateColumn()
  public createdAt: string;

  @UpdateDateColumn()
  public updatedAt: string;

  @Column('varchar')
  private password: string;

  @Column('varchar')
  private salt: string;

  public verifyPassword(retrievedPassword: string) {
    return this.password === crypto
      .createHash('sha512')
      .update(retrievedPassword + this.salt)
      .digest('hex');
  }

  public async updatePassword(retrievedPassword: string) {
    crypto.randomBytes(32, (err, buf) => {
      if (err) {
        console.error(err);
        throw new Error('Failed to generate salt.');
      } else {
        this.salt = buf.toString('hex');
        this.password = crypto
          .createHash('sha512')
          .update(retrievedPassword + this.salt)
          .digest('hex');
      }
    });
    return true;
  }

}
```

And I'm using the following code to update user profile and password:
``` TypeScript

const userRepository = getRepository(User);
const userCredentialRepository = getRepository(UserCredential);

const userCredential = await userCredentialRepository
  .findOne({
    join: {
      alias: 'user_credential',
      innerJoinAndSelect: {
        user: 'user_credential.user',
      },
    },
    where: {
      user: data.id,
    },
  })
  .catch((err) => {
    console.error(err);
    throw createError(500, 'Database operation failed.');
  });

const user = await userRepository
  .findOne({
    where: {
      id: data.id,
    },
  })
  .catch((err) => {
    console.error(err);
    throw createError(500, 'Database operation failed.');
  });

if (!user) {
  throw createError(400, 'User does not exist.');
}

if (!userCredential) {
  throw createError(400, 'User credential does not exist.');
}

// Update privilege.
user.privilege = (data.privilege === undefined) ? userCredentialInfo.user.privilege : data.privilege;

// Update password if password is given.
  if (data.password) {
    userCredential
      .updatePassword(data.password)
      .catch((err) => {
        console.error(err);
        throw createError(500, 'Update user password failed.');
      });
  }

// Update user profile.  
user.email = data.email === undefined ? user.email : data.email;
user.username = data.username === undefined ? user.username : data.username;

// Bind.
userCredential.user = user;

// Persist
await userCredentialRepository
  .persist(userCredential)
  .catch((err) => {
    console.error(err);
    throw createError(500, 'Database operation failed.');
});

```

And it fails, with:
``` sql
executing query: INSERT INTO `user`(`email`, `username`, `privilege`, `createdAt`, `updatedAt`) VALUES (?,?,?,?,?) -- PARAMETERS: ["test
@example.com","user",1,"2017-09-07T07:09:27.325Z","2017-09-07T07:09:27.325Z"]
query failed: INSERT INTO `user`(`email`, `username`, `privilege`, `createdAt`, `updatedAt`) VALUES (?,?,?,?,?) -- PARAMETERS: ["test@example.com","user",1,"2017-09-07T07:09:27.325Z","2017-09-07T07:09:27.325Z"]
error during executing query:Error: ER_DUP_ENTRY: Duplicate entry 'test@example.com' for key 'email'
```

I thought if the entry exists .persist would update the entry. I'm not sure where I'm doing wrong.

Appreciate your help.

**Comment by pleerock** [comment]
This part is wrong:

```ts
    @OneToOne(() => User, (user) => user.id, {
        primary: true,
        cascadeAll: true,
    })
```

Second argument (`(user) => user.id`) is inverse side of relation, something like user.credentials with its own one-to-one relation without join column. If you don't want to create inverse relation then you don't need to specify second argument into this decorator:

```ts
    @OneToOne(() => User, {
        primary: true,
        cascadeAll: true,
    })
```

**Comment by pleerock** [comment]
Cascades have some problems right now, so try to save your entity without cascades (take user from credentials and save user separately)

**Comment by codgician** [comment]
Thanks for replying. But if I'm saving two entities separately how can I ensure both of them would rollback if one of them fails?

**Comment by pleerock** [comment]
remove `cascadeAll: true` and re-check

**Comment by codgician** [comment]
I'm seeing the same problem after removing `cascadeAll: true`.

**Comment by pleerock** [comment]
it says so because you didn't save `user` before `credentials`. First save user, then credentials

**Comment by codgician** [comment]
Well I guess I might have figured out the problem.

When an OneToOne Column is set to primary, TypeORM would always generate "INSERT ..." sql query, resulting to a duplicate failure when the entity already exists. 

I've just tried adding another standalone primary generated column and the problem seems to be solved.

**Comment by codgician** [comment]
And when I tried to add `cascadeAll: true` back the issue went exactly like #70 .

**Comment by pleerock** [comment]
By adding primary generate column you just bring another behaviour, you arent fixing what you wanted to do.

> it says so because you didn't save user before credentials. First save user, then credentials

did you do that?

**Comment by codgician** [comment]
Sorry I might have not expressed myself clearly.

I did that. But even after I did that it still fails, displaying the same DUPLICATE ENTRY error.

**Comment by pleerock** [comment]
Can you change [this test](https://github.com/typeorm/typeorm/blob/d7f7e822c467b9edac760ed291e6d3615604eb57/test/github-issues/836/issue-836.ts#L26) and reproduce your problem here. I need minimal reproduction code of your issue

**Comment by codgician** [comment]
I created a new [repo](https://github.com/codgician/reproduction-code-836) to reproduce this issue.

You can checkout [Travis CI Page](https://travis-ci.org/codgician/reproduction-code-836#L685) for error details.

**Comment by pleerock** [comment]
[here](https://github.com/typeorm/typeorm/issues/720#issuecomment-328365922) guy had same problem and I made a fix in `0.1.0-alpha.41` version, can you please check if it works for you?

**Comment by codgician** [comment]
Perfectly solved. Appreciate your work.
